### PR TITLE
Keep AccountTypeDef to avoid DescribeAccount calls

### PR DESCRIPTION
**Pull request by iainelder**
This design improves upon my attempt in #39 because it eliminates the DescribeAccount calls completely.

### CODE REVIEW

File: botocove/cove_host_account.py [hunk 1]
```diff
@@ -128,128 +118,98 @@ def _get_boto3_sts_client(self, assuming_session: Optional[Session]) -> STSClien
...
+        }
 
-        for ou in organization_units:
-            ou_children = self._get_child_accounts(ou)
-            account_list.extend(acc["Id"] for acc in ou_children["Children"])
+        return organizational_units
 
-        return account_list
+    def _map_active_accounts(self) -> Dict[str, AccountTypeDef]:
+        pages = self.org_client.get_paginator("list_accounts").paginate()
```
Comment (connelldave) [inline botocove/cove_host_account.py:206]: I think (only had a quick pass!) this breaks functionality for non-organizations users - previously passing a list of IDs when running from any account with cross account trust to the list would run. This is particularly important for friends in the more esoteric AWS regions without AWS Orgs :)

**Comment by connelldave** [comment]
> This design improves upon my attempt in #39 because it eliminates the DescribeAccount calls completely.

Thanks for this! I was looking at #39 yesterday thinking I must help get it merged soon :)

File: botocove/cove_host_account.py (hunk 1 above)
Comment (iainelder) [reply botocove/cove_host_account.py:206]: You're surely right. I've never used botocove in that setup. Even if I pass a list of account IDs, I'm always using it in an organization.

Is it the `org_master=False` option that make it behave in non-org mode?

It would help to have some unit tests for that mode. Can we simulate it with moto?

**Comment by iainelder** [comment]
> Thanks for this! I was looking at #39 yesterday thinking I must help get it merged soon :)

Welcome back :-) I had been thinking about it for a while but I didn't want to bother you in your honeymoon. But last night I was finally inspired to try again.

File: botocove/cove_host_account.py [hunk 2]
```diff
@@ -60,7 +48,16 @@ def __init__(
             self.target_regions = regions
 
         self.provided_ignore_ids = ignore_ids
-        self.target_accounts = self._resolve_target_accounts(target_ids)
+        self.provided_target_ids = target_ids
+
+        if not org_master and self.provided_target_ids:
```
Comment (iainelder) [inline botocove/cove_host_account.py:53]: It turns out that some of the tests do use `org_master=False` but not how you might expect.

The tests don't pass any `target_ids` so botocove ends up listing the organization accounts anyway!

That seems silly and would surely fail in a non-org host account with an error along the lines of "org service not in use".

For now I want to retain that behavior, so that's why this strange if statement is here.

At least the weirdness is a  more explicit now :-)

File: botocove/cove_host_account.py (hunk 1 above)
Comment (iainelder) [reply botocove/cove_host_account.py:214]: I think with my latest commit this would be supported now. If you think it would help we could add a test that uses `org_master=False` and passes some `target_ids` to prove it.

File: botocove/cove_host_account.py (hunk 2 above)
Comment (connelldave) [reply botocove/cove_host_account.py:53]: org_master definitely has a vestigial feel to it: with this change, I think we can deprecate it and just log a warning if it's set to False as a no-op :)

The readme does sort of cover why it's there today though and the contradiction you're seeing- essentially to cop out of the problem you were having with rate limiting, but from a test point of view it's just easier to not have to mock out the describe account call... failing with an IAM error in that sense is sort of reasonable at least :)

> org_master: bool
> 
> Defaults to True. When True, will leverage the Boto3 Organizations API to list all accounts in the organization, and enrich each CoveSession with information available (Id, Arn, Name, Status, Email). Disabling this and providing your own full list of accounts may be a desirable optimisation if speed is an issue.
> 
> org_master=False means only Id will be available to CoveSession.`

File: botocove/cove_decorator.py [hunk 3]
```diff
@@ -80,3 +83,17 @@ def wrapper(*args: Any, **kwargs: Any) -> CoveOutput:
         return decorator
     else:
         return decorator(_func)
+
+
+def _validate_ids(ids: Optional[List[str]]) -> None:
```
Comment (connelldave) [inline botocove/cove_decorator.py:88]: Minor org nitpick - maybe this makes more sense within the CoveHostAccount init than in the decorator mechanism

More useful thought -  We could make these str into concrete classes earlier, so that in _list_accounts_for_resource there isn't a fail case on line 169 that can raise ValueError and we can is_instance() at that point instead of _is_regex() checks. The regex could live in that class too perhaps so we can fail fast before listing the org accounts.

It'd also make our type hints cleaner: resource_ids would be List[Union[AccountIdStr, OrgOuStr]] , account_map as Dict[AccountIdStr, AccountTypeDef]

File: botocove/cove_host_account.py [hunk 4]
```diff
@@ -60,7 +48,16 @@ def __init__(
...
+            account_map = _map_non_org_accounts(self.provided_target_ids)
+        else:
+            self.org_client = self._get_boto3_org_client(assuming_session)
```
Comment (connelldave) [inline botocove/cove_host_account.py:56]: This change does break the behaviour if someone _really_ wants to use an org account but not enumerate all accounts:
"As a niche use case, I want to use my org master, provide a specific list of accounts, but describe them individually not grab the whole org list [???]"

I'm absolutely happy to drop that support and release as 1.7.0, just want to call it out!

With the thought of just dropping the org_master bool as a supported kwarg, we could just try/except this behaviour - try and list all org accounts, if it fails fallback to the degenerate dict? 

We have these cases:
Running in org master, provides list of accounts [Will now get contents of prefetched accounts, overfetches data instead of describe_acc per acc - this seems fine]
Running in org master, does not provide accounts [normal Cove usage]
Running in standalone, provides list of accounts [won't get description - no change]
Running in standalone, provides list of OUs [will fail ungracefully? We can just log an explicit error and fail here]
Running in standalone, does not provide accounts [won't work at all]

File: botocove/cove_host_account.py [hunk 5]
```diff
@@ -128,128 +125,106 @@ def _get_boto3_sts_client(self, assuming_session: Optional[Session]) -> STSClien
...
+            descendant_accounts.update(self._list_descendant_accounts(ou))
 
-            current_ou_list: List[str] = []
+        return descendant_accounts
 
-            # current_ou_list is mutated and recursivly populated with all childs
-            self._get_all_child_ous(parent_ou, current_ou_list)
+    @lru_cache
+    def _list_accounts_for_parent(self, parent_id: str) -> Set[str]:
+        pages = self.org_client.get_paginator("list_accounts_for_parent").paginate(
```
Comment (connelldave) [inline botocove/cove_host_account.py:191]: I think a try/except block just to specifically log "you're probably not running in an org master" just to guide the failure case of non-org master passing an OU id would be a nice addition - this looks to be the place it'd surface

**Review by connelldave — COMMENTED** [review]
Looks like a great, well thought out change. I've added some notes on cases I could think of. I haven't checked the recursion logic but if the tests pass... :) That does look much cleaner too, which is excellent.

Minor note, but I've been a big fan of the .build_full_results() "secret" [weakly documented?] method on paginators, but I think your comprehension approach is my new preference - especially for clean and obvious typing!

File: botocove/cove_decorator.py (hunk 3 above)
Comment (iainelder) [reply botocove/cove_decorator.py:88]: My thinking here was to check the types as early as possible, since Python lets the user pass any old object despite what the type hint says.

I'm fine with moving it back to `CoveHostAccount` but not sure where to put it. Would you put it in `_list_accounts_for_resource` or somewhere else?

If there hadn't already been a test for this particular validation (`test_malformed_ignore_ids_type`) I would probably have ignored this part and just let duck typing happen at runtime.

To get serious about validation and domain types, would you consider using a library like [Pydantic](https://pydantic-docs.helpmanual.io/) to handle it? It seems like a big dependency to include at runtime but it's designed to handle this sort of thing.

(An aside: the validation I would really like to see is checking that `regions` is a list of strings and not a single string! So many times I have gotten errors when trying to access the regions `e`, `u`, `-`, `w`, and so on :-) )

File: botocove/cove_host_account.py (hunk 4 above)
Comment (iainelder) [reply botocove/cove_host_account.py:56]: I remember vaguely taking advantage of the behavior of the "niche" case to see a perceived speed up or increase in reliability with earlier versions of botocove (I started using it with version 1.2). I think by removing the DescribeAccount bottleneck we also remove the practical need for this use case. So it would be nice to tidy that up in version 1.7!

I like the idea of just trying to get the organization accounts and falling back to a degenerate list on some sort of error. Can we call it "graceful degredation"? :-)

If we did it like that, what would "org master" and "standalone" mean in this case? It sounds like we would assume there is an organization to query until an error is raised that proves otherwise.

File: botocove/cove_decorator.py (hunk 3 above)
Comment (connelldave) [reply botocove/cove_decorator.py:88]: I'm pretty deep into the Pydantic/FastAPI world at $job and it definitely inspired that line of thinking to have validators, but Pydantic is way too big a dependency to pull in to a library, so it's something I'm set on ruling out in this case. Having being burned by libs like `cryptography` before, I'd rather our tree is as simple as it can be.

I think we can easily get the same value out of just using stricter data types internally (but leaner than the iteration of Botocove that went a bit overboard on Dataclasses - I enjoyed learning about the dunder copy and deepcopy stuff from that adventure!) - but happy to patch that later rather than scope creep you here. 

Feel free to forget this and I'll pick up another time :)

> (An aside: the validation I would really like to see is checking that regions is a list of strings and not a single string! So many times I have gotten errors when trying to access the regions e, u, -, w, and so on :-) )

absolutely - I get bitten by this (and similarly set("str") == {"s", "t", "r"} reasonably often enough to be a bit ashamed of it :) The worst is when mypy would catch it but hasn't been run yet to point out the error and a confused debugging session ends with an "I told you so" from typecheck :D

File: botocove/cove_host_account.py (hunk 4 above)
Comment (connelldave) [reply botocove/cove_host_account.py:56]: Yeah, as per https://github.com/connelldave/botocove/pull/41#discussion_r936062666 we can just drop that terminology - either we enrich data because we can hit that API, or we gracefully degrade. If we try and then list OU contents without the ability to do that, we ungracefully explode and explain why the user is doing it wrong :)

We can even allow the super-niche case of "I don't want to describe all my accounts, here's some account and OU IDs in my org master" by just suggesting an IAM policy that denies the list org accounts, mimicing that degraded behaviour while still able to run.

Comment (iainelder) [reply botocove/cove_host_account.py:56]: > We have these cases:
> Running in org master, provides list of accounts [Will now get contents of prefetched accounts, overfetches data instead of describe_acc per acc - this seems fine]
> Running in org master, does not provide accounts [normal Cove usage]
> Running in standalone, provides list of accounts [won't get description - no change]
> Running in standalone, provides list of OUs [will fail ungracefully? We can just log an explicit error and fail here]
> Running in standalone, does not provide accounts [won't work at all]

[61f4070](https://github.com/connelldave/botocove/pull/41/commits/61f4070327c8e60d44269ce30bc97a4a7a311907) deprecates org_master and now cove beahves as if it were set True.

Surprisingly, all the tests still pass. That tells me we need new tests!

I think these cases would be good to implement as a new set of test cases. I'll work on it when I find the time. :-)

File: tests/test_new.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
+# test_new.py - I don't know what to call this yet.
```
Comment (iainelder) [inline tests/test_new.py:1]: You had been tempted in the past to discard the old tests and write new ones. I haven't referred to the existing tests yet while using this feature except to check that they pass.

I will review the existing tests to see which ones are relevant to this PR and if any of them can be updated or replaced.

File: botocove/cove_host_account.py [hunk 7]
```diff
@@ -49,18 +40,40 @@ def __init__(
 
         self.thread_workers = thread_workers
 
-        self.sts_client = self._get_boto3_sts_client(assuming_session)
-        self.org_client = self._get_boto3_org_client(assuming_session)
+        # I would pass a session around and instantiate the necessary client directly.
+        # I think we don't need special threading settings in these clients any more.
+        # CoveHostAccount is set up in a single thread.
+        # Is the client logging output considered part of the public interface?
```
Comment (iainelder) [inline botocove/cove_host_account.py:46]: Is CoveHostAccount part of the module's public interface?

I'd like the answer to be "no" so that we have more freedom to rearrange and simplify this as we wish. :-)

I'm guessing these custom boto3 client setups were required when botocove was run in two phases, but it doesn't seem necessary any more.

File: botocove/cove_host_account.py [hunk 8]
```diff
@@ -81,175 +94,229 @@ def get_cove_sessions(self) -> List[CoveSessionInformation]:
...
 
-    def _get_validated_ids(self, ids: List[str]) -> Tuple[List[str], List[str]]:
+def _host_account_is_in_organization(client: OrganizationsClient) -> bool:
+    try:
+        client.describe_organization()
+        return True
+    except ClientError as error:
+        if _organizations_service_not_in_use(error):
+            return False
+        raise error
```
Comment (connelldave) [inline botocove/cove_host_account.py:129]: Blocking problem here:

https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/organizations.html#Organizations.Client.describe_organization

> This operation can be called from any account in the organization.

https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/organizations.html#Organizations.Client.list_accounts


> This operation can be called only from the organization's management account or by a member account that is a delegated administrator for an Amazon Web Services service. 

being in an org != allowed to call list_accounts - this breaks the behaviour of using a non-org-master account within an org


Nitpick:
```suggestion
def _host_account_is_in_organization(client: OrganizationsClient) -> bool:
    try:
        client.describe_organization()
        return True
    except ClientError as error:
        if error.response["Error"]["Code"] == "AWSOrganizationsNotInUseException"
            return False
        raise error
```

think the extra func call here is a bit misleading as to something more going on

File: botocove/cove_host_account.py [hunk 9]
```diff
@@ -81,175 +94,229 @@ def get_cove_sessions(self) -> List[CoveSessionInformation]:
...
-                continue
-            raise ValueError(
-                f"provided id is neither an aws account nor an ou: {current_id}"
-            )
+def _organizations_service_not_in_use(error: ClientError) -> bool:
+    return error.response["Error"]["Code"] == "AWSOrganizationsNotInUseException"
 
-        return accounts, ous
 
-    def _get_all_accounts_by_organization_units(
```
Comment (connelldave) [inline botocove/cove_host_account.py:187]: I really like the refactor but this probably deserves it's own PR - it's just so much complexity to change all at once.

File: botocove/cove_host_account.py [hunk 10]
```diff
@@ -81,175 +94,229 @@ def get_cove_sessions(self) -> List[CoveSessionInformation]:
...
+    client: STSClient = _get_boto3_client("sts", assuming_session, thread_workers)
+    return client
+
+
+def _get_boto3_client(
+    clientname: Union[Literal["organizations"], Literal["sts"]],
+    assuming_session: Optional[Session],
+    thread_workers: int,
+) -> Any:
+    if assuming_session:
```
Comment (connelldave) [inline botocove/cove_host_account.py:312]: ah this is the reason for the odd helper function I'd forgotten :) It can almost definitely be done more elegantly but one for another time

**Review by connelldave — COMMENTED** [review]
Sorry it's taken a while to review this - it looks solid, but it's also just too big a change to actually be confident it works. We can land smaller incremental changes more easily in future :)

Happy to merge if you don't mind addressing the comments - will need a rebase as I merged the typecheck change already to be released in the next version.

File: botocove/cove_host_account.py [hunk 11]
```diff
@@ -81,175 +94,229 @@ def get_cove_sessions(self) -> List[CoveSessionInformation]:
...
+            AccountResolver(),
+            OrganizationalUnitResolver(self._org_client),
+        ]
+        for resolver in resolvers:
+            try:
+                return resolver.resolve_accounts(resource_id)
+            except ValueError:
+                continue
+        raise ValueError(
+            f"provided id is neither an aws account nor an ou: {resource_id}"
```
Comment (connelldave) [inline botocove/cove_host_account.py:207]: Think this might have been mangled somewhere? It looks like line 206 is unreachable and `AccountResolver()`  is the only resolver that'll run?

**Comment by iainelder** [comment]
Thanks for the review! I'll check you feedback in more detail when I get a moment this week.

**Comment by iainelder** [comment]
I probably won't have time now till next week to check this. Bear with me :-)

**Comment by iainelder** [comment]
I finally have time today to look at this! I'll check the blocking problem first and see if I can adress the rest of your feedback too.

File: botocove/cove_host_account.py (hunk 8 above)
Comment (iainelder) [reply botocove/cove_host_account.py:129]: > being in an org != allowed to call list_accounts - this breaks the behaviour of using a non-org-master account within an org

Yes, good catch. But I'm confused now. If you are running cove in an org, why would you run it anywhere other than the master account?

I'm having trouble understanding the use case. I can see value in setting up a "jump" account separate from the master account. Is that what you had in mind?

Comment (iainelder) [reply botocove/cove_host_account.py:129]: > Nitpick:

Agreed. It is clearer to inline the error code check in this case.

Organizations has a specific error code for the condition we care about so encapsulating the check in a function is less readable.

I had gotten into the habit of using boolean helper functions to check Boto3 client errors because some services return quite generic errors. [CloudFormation's ValidationError](https://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_DescribeStacks.html) for example) covers many conditions and to find out exactly what went wrong you need to parse the message.

File: botocove/cove_host_account.py (hunk 9 above)
Comment (iainelder) [reply botocove/cove_host_account.py:187]: Indeed, there is a lot being changed in this PR and I'm not 100% certain I haven't broken something.

My tests aren't complete, and to be honest in the current framework I'm not sure how to complete them. Moto doesn't have proper support for multiple accounts yet, so currently it's not possible to do a realistic test of the non-org modes.

File: botocove/cove_host_account.py [hunk 12]
```diff
@@ -49,18 +40,40 @@ def __init__(
...
-        self.org_client = self._get_boto3_org_client(assuming_session)
+        # I would pass a session around and instantiate the necessary client directly.
+        # I think we don't need special threading settings in these clients any more.
+        # CoveHostAccount is set up in a single thread.
+        # Is the client logging output considered part of the public interface?
+        self.sts_client = _get_boto3_sts_client(assuming_session, thread_workers)
+        self.org_client = _get_boto3_org_client(assuming_session, thread_workers)
 
-        self.host_account_id = self.sts_client.get_caller_identity()["Account"]
+        self.host_account_id = _get_host_account_id(self.sts_client)
```
[the remark this answers is not in the document]
Comment (iainelder) [reply botocove/cove_host_account.py:50]: Just a style preference: giving a domain-oriented name to small bit of functionality. It wasn't intended to change any behaviors. It's marginally clearer than the raw API call.

File: botocove/cove_host_account.py (hunk 7 above)
Comment (iainelder) [reply botocove/cove_host_account.py:46]: That's good to know! It means we could now really simplify that part of the code. But from the other comments around here I've come to the conclusion that this refactoring mixed with a big change in the functionality is too much change for one PR!

File: botocove/cove_host_account.py (hunk 11 above)
Comment (iainelder) [reply botocove/cove_host_account.py:203]: The older `test_malformed_ignore_ids` checks the the "provided id" ValueError is raised for a nonsense value `cat`. If it passes then this line must be reachable because it's the only part that generates the error.

The general pattern here is to return from first resolver that doesn't raise its own ValueError exception. If AccountResolver raises ValueError (because the resource ID is not an account ID) then we try OrganizationalUnitResolver. If that also raises ValueError (because the resource ID is not an organizational unit ID) then there are no more resolvers to try and the final ValueError is raised.

It's a bit convoluted but it was the cleanest thing I could think of to maintain the existing behavior.

**Comment by iainelder** [comment]
I think this PR has gotten a bit overwhelming for the amount of time I can dedicate to it right now. I don't see how to remove DescribeAccounts without a lot of structural changes which have knock-on effects for the non-org use cases which I don't understand very well.

My lack of understanding I think could be resolved by a solid test suite for the all the use cases that we have started to formalize in these threads.

I was about to suggest we put the change on hold until moto implements proper multi-account support, but I checked again and it just got merged 13 days ago!

https://github.com/spulec/moto/pull/5192

So that could be something worth exploring to have unit tests for the non-org modes.

But the organizations support in moto already was not complete and I imagine with the new multi-account support there may be even more use cases that it doesn't simulate well.

I'm thinking that the surest way to test some of these edge cases and use cases would be to set up a real AWS organization dedicated to testing botocove and perhaps a couple of separate non-org accounts. It wouldn't need to be a lot of accounts, just enough to test all the different use cases.

What do you think? If you think it's the right way to go then I would make a new issue to set that up and design some tests around that setup. Not sure whether it should be part of the official botocove project (as in, the "botocove project" would own the infrastructure) or whether it would be specified as a list of ingredients for a developer to set up with their own accounts and payment instrument. I'm fine with doing it either way, but I'm also conscious of not raising the bar to entry for other contributors who may not want to or be able to set up so many resources.

File: botocove/cove_host_account.py (hunk 8 above)
Comment (connelldave) [reply botocove/cove_host_account.py]: > Yes, good catch. But I'm confused now. If you are running cove in an org, why would you run it anywhere other than the master account?
> 
> I'm having trouble understanding the use case. I can see value in setting up a "jump" account separate from the master account. Is that what you had in mind?

This is something suggested as best practice: https://docs.aws.amazon.com/organizations/latest/userguide/orgs_best-practices_mgmt-acct.html#best-practices_mgmt-use - I did think about outlining in the readme which I should probably do!

A suggested solution for using Cove in automation would be to deploy a read-only role [or appopriate policy] trusting a security audit account or similar across the org.

**Comment by connelldave** [comment]
I think implementing this change in much smaller stages is going to be a lot easier to land. I'm happy to have a pass at a smaller change for review using a bunch of these ideas if you don't mind me doing that!

**Comment by connelldave** [comment]
> I think this PR has gotten a bit overwhelming for the amount of time I can dedicate to it right now. I don't see how to remove DescribeAccounts without a lot of structural changes which have knock-on effects for the non-org use cases which I don't understand very well.
> 
> My lack of understanding I think could be resolved by a solid test suite for the all the use cases that we have started to formalize in these threads.
> 
> I was about to suggest we put the change on hold until moto implements proper multi-account support, but I checked again and it just got merged 13 days ago!
> 
> [spulec/moto#5192](https://github.com/spulec/moto/pull/5192)
> 
> So that could be something worth exploring to have unit tests for the non-org modes.
> 
> But the organizations support in moto already was not complete and I imagine with the new multi-account support there may be even more use cases that it doesn't simulate well.

I'm not super worried about testing outside of what we have: Cove's complexity is only particularly high in an org context doing clever things with OU targetting and exclusion, and in the internal threading. It's an area that can be improved but I'm not sure the amount of time required is worth the value created in exhaustive coverage for the sake of it - but I don't disagree that the gaps today are making this change harder.

> I'm thinking that the surest way to test some of these edge cases and use cases would be to set up a real AWS organization dedicated to testing botocove and perhaps a couple of separate non-org accounts. It wouldn't need to be a lot of accounts, just enough to test all the different use cases.
> 
> What do you think? If you think it's the right way to go then I would make a new issue to set that up and design some tests around that setup. Not sure whether it should be part of the official botocove project (as in, the "botocove project" would own the infrastructure) or whether it would be specified as a list of ingredients for a developer to set up with their own accounts and payment instrument. I'm fine with doing it either way, but I'm also conscious of not raising the bar to entry for other contributors who may not want to or be able to set up so many resources.

I'm reticent to attach anyone's credit card to this. I might be able to do something along these lines from a work side if I pushed for more formal adoption, but I don't think we're anywhere near there yet for requiring that complexity :)

**Comment by iainelder** [comment]
> I'm happy to have a pass at a smaller change for review using a bunch of these ideas if you don't mind me doing that!

After some reflection I was preparing myself to redo the PR as a true refactor, without introducing new behavior. I think that would make the initial proposed change easier to reason about.

But I know that this month and perhaps next month my spare time will be scarce, as I too am getting married soon :-)

So feel free to push ahead and adapt the ideas in this PR into more digestable chunks and to ping me if you want a review of anything.

> I'm reticent to attach anyone's credit card to this. I might be able to do something along these lines from a work side if I pushed for more formal adoption, but I don't think we're anywhere near there yet for requiring that complexity :)

Yeah, I agree that a "corporate sponsor" would introduce a lot of complexity that doesn't necessarily give any benefits at this stage.

On my back burner of research projects I'd like to try tools like [AWS Account Controller](https://github.com/iann0036/aws-account-controller) and [Disposable Cloud Environment](https://github.com/Optum/dce) to reduce the drudgery of testing with real AWS accounts. If anything useful comes out of that, I may be able to propose a test infrastructure, but I doubt it will happen this year.

So what do you want to do with this PR? Do we keep it open till we extract the good bits? Or do we close it now?

**Comment by connelldave** [comment]
Behaviour implemented in version 1.6.3 🎉
